cockatrice/src/player.cpp

```diff
@@ -1126,46 +1126,175 @@ void Player::actCreateRelatedCard()
+                        for (int i = 0; i < cardRelation->getDefaultCount(); i++) {
+                            createCard(sourceCard, dbName);
+                        }
+                    } else {
+                        if (cardRelation->getDoesAttach()) {
+                            createAttachedCard(sourceCard, dbName);
+                        } else {
+                            createCard(sourceCard, dbName);
+                        }
+                    }
+                    break;
+                //If all are marked "Exclude", then treat the situation as if none of them are.
+                //We won't accept "garbage in, garbage out", here.
+                case 0: //else if nonExcludedRelatedCards == 0
+                    foreach (CardRelation * cardRelationAll, relatedCards) {
+                        if (!cardRelationAll->getDoesAttach() && !cardRelationAll->getIsVariable()) {
+                            dbName = dbNameFromTokenDisplayName(cardRelationAll->getName());
+                            for (int i = 0; i < cardRelationAll->getDefaultCount(); i++) {
+                                createCard(sourceCard, dbName);
+                            }
+                            tokensTypesCreated++;
+                            if(tokensTypesCreated == 1) {
+                                cardRelation = cardRelationAll;
+                            }
+                        }
+                    }
+                    break;
+                default: //else
+                    foreach (CardRelation * cardRelationNotExcluded, nonExcludedRelatedCards) {
+                        if (!cardRelationNotExcluded->getDoesAttach() && !cardRelationNotExcluded->getIsVariable()) {
+                            dbName = dbNameFromTokenDisplayName(cardRelationNotExcluded->getName());
+                            for (int i = 0; i < cardRelationNotExcluded->getDefaultCount(); i++) {
+                                createCard(sourceCard, dbName);
+                            }
+                            tokensTypesCreated++;
+                            if(tokensTypesCreated == 1) {
+                                cardRelation = cardRelationNotExcluded;
+                            }
+                        }
+                    }
+                    break;
+            }
+            break;
     }
 
+
     /*
-     * If we made a token via "Token: TokenName"
-     * then lets allow it to be created via create another
+     * If we made at least one token via "Create All Tokens"
+     * then assign the first to the "Create another" shortcut.
      */
-    if (relatedCards.length() == 1)
+    if (cardRelation != nullptr && cardRelation->getCanCreateAnother())
     {
-        CardInfo *cardInfo = db->getCard(dbNameFromTokenDisplayName(relatedCards.at(0)));
+        CardInfo *cardInfo = db->getCard(dbNameFromTokenDisplayName(cardRelation->getName()));
         setLastToken(cardInfo);
```

Comment:
I just realised that this sets the last token even if the user cancelled from a variable token creation dialog.